src/connection.rs

```diff
@@ -0,0 +1,180 @@
+                }
+                Err(e) => {
+                    error!("Error while receiving value from channel: {e}");
+                    cancel.store(true, std::sync::atomic::Ordering::Relaxed);
+                    notifier.notify_one();
+                    return;
+                }
+            }
+        }))
+    }
+
+    fn start_serf_rx(&mut self) -> Result<JoinHandle<()>, String> {
+        let notifier = Arc::clone(&self.notifier);
+        let cancel = Arc::clone(&self.cancel);
+        let dispatch = Arc::clone(&self.dispatch);
+        let mut reader = BufReader::new(self.stream.try_clone().map_err(|x| x.to_string())?);
+        Ok(std::thread::spawn(move || {
+            loop {
+                let crate::protocol::ResponseHeader { seq, error } =
+                    match rmp_serde::from_read(&mut reader) {
+                        Ok(x) => x,
+                        Err(e) => {
+                            error!("Error while reading messagepack: {e}");
+                            cancel.store(true, std::sync::atomic::Ordering::Relaxed);
+                            notifier.notify_one();
+                            return;
+                        }
+                    };
+
+                let seq_handler = {
+                    let mut dispatch = dispatch.lock().unwrap();
+                    match dispatch.map.get(&seq) {
+                        Some(v) => {
+                            if v.streaming() {
+                                if !v.stream_acked() {
+                                    continue;
+                                }
+                                v.clone()
+                            } else {
+                                dispatch.map.remove(&seq).unwrap()
+                            }
+                        }
+                        None => {
+                            // response with no handler, ignore
+                            continue;
+                        }
+                    }
+                };
+
+                let res = if error.is_empty() {
+                    Ok(SeqRead(&mut reader))
+                } else {
+                    Err(error)
+                };
+
+                seq_handler.handle(res);
+            }
+        }))
+    }
+
+    #[tokio::main(flavor = "current_thread")]
+    async fn connection_watcher(
+        dispatch: Arc<Mutex<DispatchMap>>,
+        cancel: Arc<AtomicBool>,
+        thread_notifier: Arc<Notify>,
+        stream: TcpStream,
+        serf_tx_handle: JoinHandle<()>,
+        serf_rx_handle: JoinHandle<()>,
+    ) {
+        while !cancel.load(std::sync::atomic::Ordering::Relaxed) {
+            info!("Waiting until connection threads send notification");
+            thread_notifier.notified().await;
+        }
+
+        info!("Cleaning up SERF threads");
+        drop(stream);
+
+        info!("Cleaning up existing SERF requests by responding with an error");
+        let mut lock = dispatch.lock().unwrap();
```

Comment:
Btw, `tokio::Mutex` has async lock instead of blocking, so it would look like:
```suggestion
        let mut lock = dispatch.lock().await.unwrap();
```

Reply:
Yes, however based on https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#which-kind-of-mutex-should-you-use  they recommend only using it across e.g. IO resources and to use the standard mutex whenever it is 'normal' data.